File: src/BoardMoverAD.py
from typing import Tuple

import numpy as np
from numpy.typing import NDArray
from numba import uint64, bool, njit

from BoardMover import encode_board, decode_board, encode_row, decode_row, reverse


"""
移动之后保持masked
定式计算专用，勿作他用
"""

@njit(nogil=True)
def merge_line(line: NDArray, reverse: bool = False) -> Tuple[NDArray, bool]:
    mask_new_tile = False
    if reverse:
        line = line[::-1]
    non_zero = [i for i in line if i != 0]  # 去掉所有的0
    merged = []
    skip = False
    for i in range(len(non_zero)):
        if skip:
            skip = False
            continue
        if i + 1 < len(non_zero) and non_zero[i] == non_zero[i + 1] and non_zero[i] != 32768:
            merged_value = 2 * non_zero[i]
            if merged_value >= 64:
                mask_new_tile = True
                merged_value = 32768
            merged.append(merged_value)
            skip = True
        else:
            merged.append(non_zero[i])

    # 补齐剩下的 0
    merged += [0] * (len(line) - len(merged))
    if reverse:
        merged = merged[::-1]
    return np.array(merged), mask_new_tile


@njit(nogil=True)
def calculate_all_moves() -> Tuple[NDArray, NDArray, NDArray, NDArray, NDArray]:
    # 初始化存储所有可能的行及其移动后结果差值的字典
    movel = np.empty(65536, dtype=np.uint16)
    mover = np.empty(65536, dtype=np.uint16)
    moveu = np.empty(65536, dtype=np.uint64)
    moved = np.empty(65536, dtype=np.uint64)
    mask_new_tiles = np.empty(65536, dtype=np.bool_)

    # 生成所有可能的行
    for i in range(16 ** 4):
        line = [(i // (16 ** j)) % 16 for j in range(4)]
        line = np.array([2 ** k if k else 0 for k in line])  # 把游戏中的数字转换成2的幂
        original_line = encode_row(line)  # 编码原始行为整数

        # 向左移动
        merged_linel, mask_new_tile = merge_line(line, False)
        movel[original_line] = np.uint16(encode_row(merged_linel) ^ original_line)
        mask_new_tiles[original_line] = mask_new_tile

        # 向右移动
        merged_liner, _ = merge_line(line, True)
        mover[original_line] = np.uint16(encode_row(merged_liner) ^ original_line)

    # 使用reverse函数计算向上和向下的移动差值
    for i in range(16 ** 4):
        moveu[i] = reverse(np.uint64(movel[i]))
        moved[i] = reverse(np.uint64(mover[i]))

    return movel, mover, moveu, moved, mask_new_tiles


movel, mover, moveu, moved, mask_new_tiles = calculate_all_moves()


@njit(nogil=True)
def m_move_left(board: np.uint64) -> np.uint64:
    board ^= movel[board & np.uint64(0xffff)]
    board ^= movel[board >> np.uint64(16) & np.uint64(0xffff)] << np.uint64(16)
    board ^= movel[board >> np.uint64(32) & np.uint64(0xffff)] << np.uint64(32)
    board ^= movel[board >> np.uint64(48) & np.uint64(0xffff)] << np.uint64(48)
    return board


@njit(nogil=True)
def m_move_left2(board: np.uint64) -> Tuple[np.uint64, bool]:
    mnt = False
    mnt |= mask_new_tiles[board & np.uint64(0xffff)]
    board ^= movel[board & np.uint64(0xffff)]
    mnt |= mask_new_tiles[board >> np.uint64(16) & np.uint64(0xffff)]
    board ^= movel[board >> np.uint64(16) & np.uint64(0xffff)] << np.uint64(16)
    mnt |= mask_new_tiles[board >> np.uint64(32) & np.uint64(0xffff)]
    board ^= movel[board >> np.uint64(32) & np.uint64(0xffff)] << np.uint64(32)
    mnt |= mask_new_tiles[board >> np.uint64(48) & np.uint64(0xffff)]
    board ^= movel[board >> np.uint64(48) & np.uint64(0xffff)] << np.uint64(48)
    return board, mnt


@njit(nogil=True)
def m_move_right(board: np.uint64) -> np.uint64:
    board ^= mover[board & np.uint64(0xffff)]
    board ^= mover[board >> np.uint64(16) & np.uint64(0xffff)] << np.uint64(16)
    board ^= mover[board >> np.uint64(32) & np.uint64(0xffff)] << np.uint64(32)
    board ^= mover[board >> np.uint64(48) & np.uint64(0xffff)] << np.uint64(48)
    return board


@njit(nogil=True)
def m_move_up(board: np.uint64, board2: np.uint64) -> np.uint64:
    board ^= moveu[board2 & np.uint64(0xffff)]
    board ^= moveu[board2 >> np.uint64(16) & np.uint64(0xffff)] << np.uint64(4)
    board ^= moveu[board2 >> np.uint64(32) & np.uint64(0xffff)] << np.uint64(8)
    board ^= moveu[board2 >> np.uint64(48) & np.uint64(0xffff)] << np.uint64(12)
    return board


@njit(nogil=True)
def m_move_up2(board: np.uint64, board2: np.uint64) -> Tuple[np.uint64, bool]:
    mnt = False
    mnt |= mask_new_tiles[board2 & np.uint64(0xffff)]
    board ^= moveu[board2 & np.uint64(0xffff)]
    mnt |= mask_new_tiles[board2 >> np.uint64(16) & np.uint64(0xffff)]
    board ^= moveu[board2 >> np.uint64(16) & np.uint64(0xffff)] << np.uint64(4)
    mnt |= mask_new_tiles[board2 >> np.uint64(32) & np.uint64(0xffff)]
    board ^= moveu[board2 >> np.uint64(32) & np.uint64(0xffff)] << np.uint64(8)
    mnt |= mask_new_tiles[board2 >> np.uint64(48) & np.uint64(0xffff)]
    board ^= moveu[board2 >> np.uint64(48) & np.uint64(0xffff)] << np.uint64(12)
    return board, mnt


@njit(nogil=True)
def m_move_down(board: np.uint64, board2: np.uint64) -> np.uint64:
    board ^= moved[board2 & np.uint64(0xffff)]
    board ^= moved[board2 >> np.uint64(16) & np.uint64(0xffff)] << np.uint64(4)
    board ^= moved[board2 >> np.uint64(32) & np.uint64(0xffff)] << np.uint64(8)
    board ^= moved[board2 >> np.uint64(48) & np.uint64(0xffff)] << np.uint64(12)
    return board


@njit(nogil=True)
def m_move_board(board: np.uint64, direction: int) -> np.uint64:
    if direction == 1:
        return m_move_left(board)
    elif direction == 2:
        return m_move_right(board)
    elif direction == 3:
        board2 = reverse(board)
        return m_move_up(board, board2)
    elif direction == 4:
        board2 = reverse(board)
        return m_move_down(board, board2)
    else:
        print(f'bad direction input:{direction}')
        return board


@njit(nogil=True)
def m_move_board2(board: np.uint64, board2: np.uint64, direction: int) -> np.uint64:
    if direction == 1:
        return m_move_left(board)
    elif direction == 2:
        return m_move_right(board)
    elif direction == 3:
        return m_move_up(board, board2)
    else:  # direction == 4:
        return m_move_down(board, board2)


@njit(nogil=True)
def m_move_all_dir(board: np.uint64
                   ) -> Tuple[Tuple[uint64, bool], Tuple[uint64, bool], Tuple[uint64, bool], Tuple[uint64, bool]]:
    board = np.uint64(board)
    board2 = reverse(board)
    md = m_move_down(board, board2)
    mr = m_move_right(board)
    ml, mnt_h = m_move_left2(board)
    mu, mnt_v = m_move_up2(board, board2)
    return (ml, mnt_h), (mr, mnt_h), (mu, mnt_v), (md, mnt_v)


@njit(nogil=True)
def m_move_all_dir2(board: np.uint64, board2: np.uint64
                    ) -> Tuple[Tuple[uint64, bool], Tuple[uint64, bool], Tuple[uint64, bool], Tuple[uint64, bool]]:
    md = m_move_down(board, board2)
    mr = m_move_right(board)
    ml, mnt_h = m_move_left2(board)
    mu, mnt_v = m_move_up2(board, board2)
    return (ml, mnt_h), (mr, mnt_h), (mu, mnt_v), (md, mnt_v)


if __name__ == "__main__":
    print(np.sum(mask_new_tiles))
    b = np.uint64(0x000101020aa37bcf)
    b_r = np.uint64(reverse(b))
    print(decode_board(b))
    print()
    for nt, _mnt in m_move_all_dir(b):
        print(decode_board(nt))
        print(_mnt)

    for nt, _mnt in m_move_all_dir2(b, b_r):
        print(decode_board(nt))
        print(_mnt)

File: native_core/src/BoardMoverAD.cpp
#include "BoardMoverAD.h"

#include <algorithm>
#include <vector>

namespace FormationAD {

namespace {

struct ADRowMove {
    uint16_t movel = 0;
    uint16_t mover = 0;
};

std::array<ADRowMove, 65536> g_row_table{};
std::array<uint64_t, 1024> g_mask_new_tile_bits{};

inline void set_mask_new_tile(uint32_t row) {
    g_mask_new_tile_bits[row >> 6U] |= (1ULL << (row & 63U));
}

inline bool has_mask_new_tile(uint64_t row) {
    return ((g_mask_new_tile_bits[row >> 6U] >> (row & 63U)) & 1ULL) != 0ULL;
}

struct ADRowTableInitializer {
    ADRowTableInitializer() {
        for (uint32_t i = 0; i < 65536; ++i) {
            bool mask_new_tile = false;
            uint16_t left = merge_line(static_cast<uint16_t>(i), false, mask_new_tile);

            bool ignored = false;
            uint16_t right = merge_line(static_cast<uint16_t>(i), true, ignored);
            g_row_table[i] = ADRowMove{
                static_cast<uint16_t>(left ^ static_cast<uint16_t>(i)),
                static_cast<uint16_t>(right ^ static_cast<uint16_t>(i))
            };
            if (mask_new_tile) {
                set_mask_new_tile(i);
            }
        }
    }
};

ADRowTableInitializer g_initializer;

} // namespace

uint16_t merge_line(uint16_t row, bool reverse_line, bool &mask_new_tile) {
    uint32_t line[4];
    decode_row(row, line);
    if (reverse_line) {
        std::reverse(std::begin(line), std::end(line));
    }

    std::vector<uint32_t> non_zero;
    non_zero.reserve(4);
    for (uint32_t value : line) {
        if (value != 0) {
            non_zero.push_back(value);
        }
    }

    uint32_t merged[4] = {0, 0, 0, 0};
    size_t merged_count = 0;
    bool skip = false;
    for (size_t i = 0; i < non_zero.size(); ++i) {
        if (skip) {
            skip = false;
            continue;
        }
        if (i + 1 < non_zero.size() && non_zero[i] == non_zero[i + 1] && non_zero[i] != 32768U) {
            uint32_t merged_value = non_zero[i] * 2U;
            if (merged_value >= 64U) {
                mask_new_tile = true;
                merged_value = 32768U;
            }
            merged[merged_count++] = merged_value;
            skip = true;
        } else {
            merged[merged_count++] = non_zero[i];
        }
    }

    if (reverse_line) {
        std::reverse(std::begin(merged), std::end(merged));
    }
    return encode_row(merged);
}

uint64_t m_move_left(uint64_t board) {
    board ^= static_cast<uint64_t>(g_row_table[board & 0xFFFFULL].movel);
    board ^= static_cast<uint64_t>(g_row_table[(board >> 16U) & 0xFFFFULL].movel) << 16U;
    board ^= static_cast<uint64_t>(g_row_table[(board >> 32U) & 0xFFFFULL].movel) << 32U;
    board ^= static_cast<uint64_t>(g_row_table[(board >> 48U) & 0xFFFFULL].movel) << 48U;
    return board;
}

std::pair<uint64_t, bool> m_move_left2(uint64_t board) {
    const uint64_t r0 = board & 0xFFFFULL;
    const uint64_t r1 = (board >> 16U) & 0xFFFFULL;
    const uint64_t r2 = (board >> 32U) & 0xFFFFULL;
    const uint64_t r3 = (board >> 48U) & 0xFFFFULL;
    const ADRowMove &e0 = g_row_table[r0];
    const ADRowMove &e1 = g_row_table[r1];
    const ADRowMove &e2 = g_row_table[r2];
    const ADRowMove &e3 = g_row_table[r3];
    const bool mnt = has_mask_new_tile(r0) || has_mask_new_tile(r1) || has_mask_new_tile(r2) || has_mask_new_tile(r3);
    board ^= static_cast<uint64_t>(e0.movel);
    board ^= static_cast<uint64_t>(e1.movel) << 16U;
    board ^= static_cast<uint64_t>(e2.movel) << 32U;
    board ^= static_cast<uint64_t>(e3.movel) << 48U;
    return {board, mnt};
}

uint64_t m_move_right(uint64_t board) {
    board ^= static_cast<uint64_t>(g_row_table[board & 0xFFFFULL].mover);
    board ^= static_cast<uint64_t>(g_row_table[(board >> 16U) & 0xFFFFULL].mover) << 16U;
    board ^= static_cast<uint64_t>(g_row_table[(board >> 32U) & 0xFFFFULL].mover) << 32U;
    board ^= static_cast<uint64_t>(g_row_table[(board >> 48U) & 0xFFFFULL].mover) << 48U;
    return board;
}

uint64_t m_move_up(uint64_t board, uint64_t board_rev) {
    (void) board;
    return reverse_board(m_move_left(board_rev));
}

std::pair<uint64_t, bool> m_move_up2(uint64_t board, uint64_t board_rev) {
    (void) board;
    auto [moved_rev, mnt] = m_move_left2(board_rev);
    return {reverse_board(moved_rev), mnt};
}

uint64_t m_move_down(uint64_t board, uint64_t board_rev) {
    (void) board;
    return reverse_board(m_move_right(board_rev));
}

uint64_t m_move_board(uint64_t board, int direction) {
    if (direction == 1) {
        return m_move_left(board);
    }
    if (direction == 2) {
        return m_move_right(board);
    }
    if (direction == 3) {
        uint64_t board_rev = reverse(board);
        return m_move_up(board, board_rev);
    }
    if (direction == 4) {
        uint64_t board_rev = reverse(board);
        return m_move_down(board, board_rev);
    }
    return board;
}

uint64_t m_move_board2(uint64_t board, uint64_t board_rev, int direction) {
    if (direction == 1) {
        return m_move_left(board);
    }
    if (direction == 2) {
        return m_move_right(board);
    }
    if (direction == 3) {
        return m_move_up(board, board_rev);
    }
    return m_move_down(board, board_rev);
}

std::array<ADMoveResult, 4> m_move_all_dir(uint64_t board) {
    uint64_t board_rev = reverse(board);
    uint64_t md = m_move_down(board, board_rev);
    uint64_t mr = m_move_right(board);
    auto [ml, mnt_h] = m_move_left2(board);
    auto [mu, mnt_v] = m_move_up2(board, board_rev);
    return {ADMoveResult{ml, mnt_h}, ADMoveResult{mr, mnt_h}, ADMoveResult{mu, mnt_v}, ADMoveResult{md, mnt_v}};
}

std::array<ADMoveResult, 4> m_move_all_dir2(uint64_t board, uint64_t board_rev) {
    uint64_t md = m_move_down(board, board_rev);
    uint64_t mr = m_move_right(board);
    auto [ml, mnt_h] = m_move_left2(board);
    auto [mu, mnt_v] = m_move_up2(board, board_rev);
    return {ADMoveResult{ml, mnt_h}, ADMoveResult{mr, mnt_h}, ADMoveResult{mu, mnt_v}, ADMoveResult{md, mnt_v}};
}

} // namespace FormationAD
